File: src/oca/java/src/org/opennebula/client/Pool.java
/*******************************************************************************
 * Copyright 2002-2012, OpenNebula Project Leads (OpenNebula.org)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client;

import java.io.ByteArrayInputStream;
import java.io.IOException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import org.w3c.dom.Node;
import org.w3c.dom.Element;
import org.xml.sax.SAXException;

/**
 * Represents a generic OpenNebula Pool in XML format
 * and provides the basic functionality to handle the Pool elements.
 */
public abstract class Pool{

    protected Client client;

    protected String   elementName;
    protected String   infoMethod;
    protected NodeList poolElements;

    /**
     * All resources in the pool
     */
    public final static int ALL        = -2;

    /**
     * Connected user's resources
     */
    public final static int MINE       = -3;

    /**
     * Connected user's resources, and the ones in his group
     */
    public final static int MINE_GROUP = -1;

    /**
     * Protected constructor, to be called from subclasses.
     *
     * @param elementName Name of the PoolElement's xml element
     * @param client XML-RPC client which will handle calls
     * @param infoMethod XML-RPC info method for the subclass Pool
     */
    protected Pool(String elementName, Client client, String infoMethod)
    {
        this.elementName = elementName;
        this.infoMethod  = infoMethod;
        this.client      = client;
    }

    /**
     * The factory method returns a suitable PoolElement object from
     * an XML node. Each Pool must implement the corresponding factory method.
     *
     * @param node XML Dom node to build the PoolElement from
     * @return The corresponding PoolElement
     */
    public abstract PoolElement factory(Node node);

    /***************************************************************************
     * Info methods
     **************************************************************************/

    protected static OneResponse info(Client client, String infoMethod)
    {
        return xmlrpcInfo(client, infoMethod);
    }

    protected static OneResponse info(Client client, String infoMethod,
            int filter, int startId, int endId)
    {
        return xmlrpcInfo(client, infoMethod, filter, startId, endId);
    }

    protected static OneResponse infoAll(Client client, String infoMethod)
    {
        return xmlrpcInfo(client, infoMethod, ALL, -1, -1);
    }

    protected static OneResponse infoMine(Client client, String infoMethod)
    {
        return xmlrpcInfo(client, infoMethod, MINE, -1, -1);
    }

    protected static OneResponse infoGroup(Client client, String infoMethod)
    {
        return xmlrpcInfo(client, infoMethod, MINE_GROUP, -1, -1);
    }

    private static OneResponse xmlrpcInfo(Client client, String infoMethod, Object...args)
    {
        return client.call(infoMethod, args);
    }

    protected OneResponse info()
    {
        OneResponse response = info(client, infoMethod);
        processInfo(response);
        return response;
    }

    protected OneResponse infoAll()
    {
        OneResponse response = infoAll(client, infoMethod);
        processInfo(response);
        return response;
    }

    protected OneResponse infoMine()
    {
        OneResponse response = infoMine(client, infoMethod);
        processInfo(response);
        return response;
    }

    protected OneResponse infoGroup()
    {
        OneResponse response = infoGroup(client, infoMethod);
        processInfo(response);
        return response;
    }

    protected OneResponse info(int filter, int startId, int endId)
    {
        OneResponse response = info(client, infoMethod, filter, startId, endId);
        processInfo(response);
        return response;
    }

    /**
     * After a *pool.info call, this method builds the internal xml
     * representation of the pool.
     * @param info The XML-RPC *pool.info response
     */
    public void processInfo(OneResponse info)
    {
        if (info.isError())
        {
            return;
        }

        try
        {
            DocumentBuilder builder;
            Document        doc;
            Element         xml;

            builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
            doc = builder.parse(
                new ByteArrayInputStream(info.getMessage().getBytes()));
            xml = doc.getDocumentElement();

            poolElements = xml.getElementsByTagName(elementName);
        }
        catch (ParserConfigurationException e) {}
        catch (SAXException e) {}
        catch (IOException e) {}
    }

    /**
     * Returns the indexth element in the pool. If index is greater than or
     * equal to the number of elements in the pool, this returns null.
     *
     * @param index Index of the element.
     * @return The element at the indexth position in the pool, or
     * null if that is not a valid index.
     */
    public PoolElement item(int index)
    {
        PoolElement theElement = null;

        if (poolElements != null)
        {
            Node node = poolElements.item(index);

            if (node != null)
            {
                theElement = factory(node);
            }
        }

        return theElement;
    }

    /**
     * Returns the element with the given Id from the pool. If it is not found,
     * then returns null. The method {@link #info()} must be called before.
     *
     * @param id of the element to retrieve
     * @return The element with the given Id, or null if it was not found.
     */
    protected PoolElement getById(int id)
    {
        // TODO: Use xpath to find the element /<elementName>/ID

        PoolElement theElement = null;
        PoolElement tmpElement = null;

        for( int i = 0; i < getLength(); i++ )
        {
            tmpElement = item(i);
            if( tmpElement.id() == id )
            {
                theElement = tmpElement;
                break;
            }
        }

        return theElement;
    }

    /**
     * The number of elements in the pool.
     * @return The number of elements in the pool.
     */
    public int getLength()
    {
        return poolElements == null ? 0 : poolElements.getLength();
    }
}

File: src/oca/ruby/opennebula/pool.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

require 'opennebula/xml_utils'

module OpenNebula

    # The Pool class represents a generic OpenNebula Pool in XML format
    # and provides the basic functionality to handle the Pool elements
    class Pool < XMLPool
        include Enumerable
        alias_method :each_with_xpath, :each

        attr_reader :pool_name
        attr_reader :element_name

        PAGINATED_POOLS=%w{VM_POOL IMAGE_POOL TEMPLATE_POOL VN_POOL
                           SECGROUP_POOL DOCUMENT_POOL}

    protected
        #pool:: _String_ XML name of the root element
        #element:: _String_ XML name of the Pool elements
        #client::  _Client_ represents a XML-RPC connection
        def initialize(pool,element,client)
            super(nil)

            @pool_name    = pool.upcase
            @element_name = element.upcase

            @client = client
        end

        # Default Factory Method for the Pools. The factory method returns an
        # suitable PoolElement object. Each Pool MUST implement the
        # corresponding factory method
        # element_xml:: _XML_ XML element describing the pool element
        # [return] a PoolElement object
        def factory(element_xml)
            OpenNebula::PoolElement.new(element_xml,client)
        end

        #######################################################################
        # Common XML-RPC Methods for all the Pool Types
        #######################################################################

        #Gets the pool without any filter. Host, Group and User Pools
        # xml_method:: _String_ the name of the XML-RPC method
        def info(xml_method)
            return xmlrpc_info(xml_method)
        end

        alias_method :info!, :info

        def info_extended(xml_method)
            return xmlrpc_info(xml_info)
        end

        def info_all(xml_method, *args)
            return xmlrpc_info(xml_method, INFO_ALL, -1, -1, *args)
        end

        def info_mine(xml_method, *args)
            return xmlrpc_info(xml_method, INFO_MINE, -1, -1, *args)
        end

        def info_group(xml_method, *args)
            return xmlrpc_info(xml_method, INFO_GROUP, -1, -1, *args)
        end

        def info_primary_group(xml_method, *args)
            return xmlrpc_info(xml_method, INFO_PRIMARY_GROUP, -1, -1, *args)
        end

        def info_filter(xml_method, who, start_id, end_id, *args)
            return xmlrpc_info(xml_method, who, start_id, end_id, *args)
        end

        # Retrieves the monitoring data for all the Objects in the pool
        #
        # @param [String] xml_method xml-rcp method
        # @param [Array<String>] xpath_expressions Elements to retrieve.
        # @param args arguemnts for the xml_method call
        #
        # @return [Hash<String, <Hash<String, Array<Array<int>>>>>,
        #   OpenNebula::Error] The first level hash uses the Object ID as keys,
        #   and as value a Hash with the requested xpath expressions,
        #   and an Array of 'timestamp, value'.
        def monitoring(xml_method, xpaths, *args)

            rc = @client.call(xml_method, *args)

            if ( OpenNebula.is_error?(rc) )
                return rc
            end

            xmldoc = XMLElement.new
            xmldoc.initialize_xml(rc, 'MONITORING_DATA')

            hash = {}

            # Get all existing Object IDs
            ids = xmldoc.retrieve_elements('/MONITORING_DATA/MONITORING/ID')

            if ids.nil?
                return hash
            else
                ids.uniq!
            end

            ids.each { |id|
                hash[id] = OpenNebula.process_monitoring(xmldoc, id, xpaths)
            }

            return hash
        end

        def monitoring_last(xml_method, *args)
            rc = @client.call(xml_method, *args)

            if OpenNebula.is_error?(rc)
                return rc
            end

            xmldoc = XMLElement.new
            xmldoc.initialize_xml(rc, 'MONITORING_DATA')

            xmldoc.to_hash
        end

    private
        # Calls to the corresponding info method to retreive the pool
        # representation in XML format
        # xml_method:: _String_ the name of the XML-RPC method
        # args:: _Array_ with additional arguments for the info call
        # [return] nil in case of success or an Error object
        def xmlrpc_info(xml_method, *args)
            rc = @client.call(xml_method, *args)

            if !OpenNebula.is_error?(rc)
                initialize_xml(rc, @pool_name)

                rc = nil
            end

            return rc
        end

    public
        # Constants for info queries (include/RequestManagerPoolInfoFilter.h)
        INFO_GROUP = -1
        INFO_ALL   = -2
        INFO_MINE  = -3
        INFO_PRIMARY_GROUP = -4

        # Iterates over every PoolElement in the Pool and calls the block with a
        # a PoolElement obtained calling the factory method
        # block:: _Block_
        def each(&block)
            each_element(block) if @xml
        end

        # DO NOT USE - ONLY REXML BACKEND
        def to_str
            str = ""
            REXML::Formatters::Pretty.new(1).write(@xml,str)

            return str
        end

        # Gets a hash from a pool
        #
        # size:: nil => default page size
        #        < 2 => not paginated
        #        >=2 => page size
        #
        # The default page size can be changed with the environment variable
        # ONE_POOL_PAGE_SIZE. Any value > 2 will set a page size, a non
        # numeric value disables pagination.
        def get_hash(size=nil)
            allow_paginated = PAGINATED_POOLS.include?(@pool_name)

            if OpenNebula.pool_page_size && allow_paginated &&
                    ( ( size && size >= 2 ) || !size )

                size = OpenNebula.pool_page_size if !size
                hash = info_paginated(size)

                return hash if OpenNebula.is_error?(hash)

                { @pool_name => { @element_name => hash } }
            else
                rc = info
                return rc if OpenNebula.is_error?(rc)

                to_hash
            end
        end

        # Gets a pool in hash form using pagination
        #
        # size:: _Integer_ size of each page
        def info_paginated(size)
            array   = Array.new
            current = 0

            parser = ParsePoolSax.new(@pool_name, @element_name)

            while true
                a = @client.call("#{@pool_name.delete('_').downcase}.info",
                        @user_id, current, -size, -1)

                return a if OpenNebula.is_error?(a)

                a_array=parser.parse(a)

                array   += a_array
                current += size

                break if !a || a_array.length<size
            end

            array.compact!
            array = nil if array.length == 0

            array
        end

        # Gets a hash from a info page from pool
        # size:: nil => default page size
        #        > 0 => page size
        # current first element of the page
        # extended true to get extended information
        # state state of the objects
        # hash:: return page as a hash
        def get_page(size, current, extended = false, state = -1)
            rc      = nil
            state ||= -1

            if PAGINATED_POOLS.include?(@pool_name)
                pool_name = @pool_name.delete('_').downcase

                if extended && pool_name == "vmpool"
                    method = "#{pool_name}.infoextended"
                else
                    method = "#{pool_name}.info"
                end

                size = OpenNebula.pool_page_size if (!size || size == 0)
                rc   = @client.call(method, @user_id, current, -size, state)

                return rc if OpenNebula.is_error?(rc)

                initialize_xml(rc, @pool_name)
            else
                rc = info
            end

            return rc
        end

        # Iterates over pool page
        def loop_page(size, state, extended)
            current = 0
            element = @pool_name.split('_')[0]
            page    = OpenNebula::XMLElement.new

            loop do
                rc = get_page(size, current, extended, state)

                break rc if OpenNebula.is_error?(rc)

                page.initialize_xml(rc, @pool_name)

                break if page["//#{element}"].nil?

                current += yield(element, page)
            end
        end

        # Iterates over pool pages
        # size:: nil => default page size
        #        > 0 => page size
        # state state of objects
        def each_page(size, state = -1, extended = false)
            loop_page(size, state, extended) do |element, page|
                page.each("//#{element}") do |obj|
                    yield(obj)
                end

                size
            end
        end

        # Iterates over pool pages to delete them
        # size:: nil => default page size
        #        > 0 => page size
        # state state of objects
        def each_page_delete(size, state = -1, extended = false)
            loop_page(size, state, extended) do |element, page|
                no_deleted = 0

                page.each("//#{element}") do |obj|
                    no_deleted += 1 if !yield(obj)
                end

                no_deleted
            end
        end

        # Return true if pool is paginated
        def is_paginated?
            PAGINATED_POOLS.include?(@pool_name)
        end

    end

end
